Spending process: Cosigning server + Coordinator communication, and Unvault + Spend broadcast

darosior opened the pull request:
This implements the spending process on top of #125 .

We add a `spend_transactions` table to the database in order to keep track of PSBTs. A bridge `spend_inputs` table is added as well to join with the Unvault transaction in the `presigned_transactions` table.

This allows to offer to the caller a way to:
1. Create a Spend transaction (previous work, already in master)
2. Store or update a Spend transaction (`updatespendtx`)
3. Delete a stored Spend transaction (`delspendtx`)
4. List the stored Spend transaction along with the vault deposit outpoints they ultimately spend (`listspendtxs`)
5. Trigger the Spending Process by
5.a Checking all managers signed
5.b Requesting the signatures from the cosigning servers
5.c Announcing the Spend transaction to the Coordinator
5.d Broadcasting the Unvault transaction(s)
5.e Later broadcasting the Spend transaction

We also integrate the Cosigning Server in the functional test framework: https://github.com/revault/cosignerd .

This does not contain proper state tracking which will be done in a followup.

Fixes #118

darosior commented:
Added a drive-by (hopefully) fix for the second item of #118

darosior commented:
The last commit was added as a prep work of a more extensive blackbox test of the Spend process, which will finally not happen in this PR. Leaving it though as it's tightly related to the Spend process.

edouardparis commented on .gitmodules:
```diff
@@ -1,3 +1,6 @@
 [submodule "tests/servers/coordinatord"]
 	path = tests/servers/coordinatord
 	url = https://github.com/revault/coordinatord
+[submodule "tests/servers/cosignerd"]
```
update tests/README.md https://github.com/revault/revaultd/tree/master/tests#with-the-servers

darosior replied:
Good catch, thanks!

darosior commented:
We should not be able to delete a Spend marked as `broadcast = 1`

darosior commented:
Edouard feedback: `updatespendtx` should not take outpoints

danielabrozzoni commented on src/daemon/control.rs:
```diff
@@ -1050,6 +1216,208 @@ pub fn handle_rpc_messages(
+
+                // Does the Spend psbt we are passed actually correspond to the outpoint array?
+                fn all_outpoints_match(
+                    spend_inputs: &Vec<revault_tx::bitcoin::TxIn>,
```
From `cargo clippy`:
```
warning: writing `&Vec<_>` instead of `&[_]` involves one more reference and cannot be used with non-Vec-based slices.
    --> src/daemon/control.rs:1288:35
     |
1288 |                     spend_inputs: &Vec<revault_tx::bitcoin::TxIn>,
     |                                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: change this to: `&[revault_tx::bitcoin::TxIn]`
     |
     = note: `#[warn(clippy::ptr_arg)]` on by default
     = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#ptr_arg
```
(same for `db_unvaults`)

darosior commented:
- Removed the `outpoints` from `updatespendtx` arguments
- Updated the tests doc
- Added some wannabe-edgecase-testing test cases, new suggestions of not a scenario not currently covered **very welcome**!

danielabrozzoni commented on src/daemon/database/actions.rs:
```diff
@@ -487,7 +487,10 @@ pub fn db_update_presigned_tx(
 pub fn db_insert_spend(
     db_path: &PathBuf,
-    unvault_txs: &[&DbTransaction],
+    // FIXME: Rust newbie: i *don't* need to be moving this. So i want to take &[&T] but i can't
```
You're not moving the transactions, are you?
https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=1be844b0305ec94fe385c8efffb5fb74

darosior replied:
Hmm it does move as for example
```patch
diff --git a/src/daemon/database/actions.rs b/src/daemon/database/actions.rs
index 9b370dd..c1e436d 100644
--- a/src/daemon/database/actions.rs
+++ b/src/daemon/database/actions.rs
@@ -1055,7 +1055,7 @@ mod test {
             db_vault_by_unvault_txid(&db_path, &spend_tx_inputs[0].previous_output.txid)
                 .unwrap()
                 .unwrap();
-        db_insert_spend(&db_path, &[db_unvault.clone()], &spend_tx).unwrap();
+        db_insert_spend(&db_path, &[db_unvault], &spend_tx).unwrap();
         let spend_txid = spend_tx.inner_tx().global.unsigned_tx.txid();
         assert_eq!(
             db_list_spends(&db_path).unwrap().get(&spend_txid),
```
Does not compile (with a nice use-after-move error message)

darosior replied:
You were right, i'm actually moving it by creating an array each time! This compiles
```diff
diff --git a/src/daemon/database/actions.rs b/src/daemon/database/actions.rs
index 9b370dd..2c9a3da 100644
--- a/src/daemon/database/actions.rs
+++ b/src/daemon/database/actions.rs
@@ -1055,7 +1055,8 @@ mod test {
             db_vault_by_unvault_txid(&db_path, &spend_tx_inputs[0].previous_output.txid)
                 .unwrap()
                 .unwrap();
-        db_insert_spend(&db_path, &[db_unvault.clone()], &spend_tx).unwrap();
+        let arr = [db_unvault.clone()];
+        db_insert_spend(&db_path, &arr, &spend_tx).unwrap();
         let spend_txid = spend_tx.inner_tx().global.unsigned_tx.txid();
         assert_eq!(
             db_list_spends(&db_path).unwrap().get(&spend_txid),
@@ -1078,7 +1079,7 @@ mod test {
         // And this works with multiple unvaults too
 
         // Re-insert the previous one so we have many references to the first Unvault
-        db_insert_spend(&db_path, &[db_unvault.clone()], &spend_tx).unwrap();
+        db_insert_spend(&db_path, &arr, &spend_tx).unwrap();
```

darosior replied:
Removed the comment.

danielabrozzoni commented on src/daemon/control.rs:
```diff
@@ -1103,6 +1103,20 @@ pub fn handle_rpc_messages(
+
+                let mut spend_tx_map = db_list_spends(&db_path)?;
+                let mut listspend_entries = Vec::with_capacity(spend_tx_map.len());
+                for (_, (psbt, deposit_outpoints)) in spend_tx_map.drain() {
```
nit: Do you really need the `drain()` here?
```
--- a/src/daemon/control.rs
+++ b/src/daemon/control.rs
@@ -1284,9 +1284,9 @@ pub fn handle_rpc_messages(
             RpcMessageIn::ListSpendTxs(response_tx) => {
                 let db_path = revaultd.read().unwrap().db_file();
 
-                let mut spend_tx_map = db_list_spends(&db_path)?;
+                let spend_tx_map = db_list_spends(&db_path)?;
                 let mut listspend_entries = Vec::with_capacity(spend_tx_map.len());
-                for (_, (psbt, deposit_outpoints)) in spend_tx_map.drain() {
+                for (_, (psbt, deposit_outpoints)) in spend_tx_map {
                     listspend_entries.push(ListSpendEntry {
```

darosior replied:
I found it somewhat more explicit, but i can change it

darosior replied:
Ok went for it to remove the `mut`, thanks!

darosior commented:
Made the `test_large_spends` test case sign-and-wait in parallel. Resulted in a factor 2 speedup:
```diff
diff --git a/tests/test_misc.py b/tests/test_misc.py
index cfad5ea..a9a7d64 100644
--- a/tests/test_misc.py
+++ b/tests/test_misc.py
@@ -1142,7 +1142,7 @@ def test_spends_conflicting(revault_network, bitcoind):
 # FIXME: exchange of signatures with the cosigning server gets too large too quickly
 # See https://github.com/revault/practical-revault/issues/81
 @pytest.mark.skipif(not POSTGRES_IS_SETUP, reason="Needs Postgres for servers db")
-def test_large_spends(revault_network, bitcoind):
+def test_large_spends(revault_network, bitcoind, executor):
     CSV = 2016  # 2 weeks :tm:
     revault_network.deploy(9, 4, csv=CSV)
     man = revault_network.man_wallets[0]
@@ -1150,11 +1150,23 @@ def test_large_spends(revault_network, bitcoind):
     deposits = []
     deriv_indexes = []
     total_amount = 0
+    revtx_futures = []
     for _ in range(15):
         amount = random.randint(5, 5000) / 100
         vault = revault_network.fund(amount)
-        revault_network.secure_vault(vault)
-        revault_network.activate_vault(vault)
+        revtx_futures.append(
+            (vault, executor.submit(revault_network.secure_vault, vault))
+        )
+
+    unvaulttx_futures = []
+    for (vault, fut) in revtx_futures:
+        fut.result(TIMEOUT)
+        unvaulttx_futures.append(
+            (vault, executor.submit(revault_network.activate_vault, vault))
+        )
+
+    for (vault, fut) in unvaulttx_futures:
+        fut.result(TIMEOUT)
         deposits.append(f"{vault['txid']}:{vault['vout']}")
         deriv_indexes.append(vault["derivation_index"])
         total_amount += vault["amount"]
 ```

edouardparis commented on src/daemon/database/schema.rs:
```diff
@@ -68,20 +69,47 @@ CREATE TABLE presigned_transactions (
+    id INTEGER PRIMARY KEY NOT NULL,
+    psbt BLOB UNIQUE NOT NULL,
+    txid BLOB UNIQUE NOT NULL,
+    broadcast BOOLEAN NOT NULL CHECK (broadcast IN (0,1))
```
Maybe move to a ternary status: 0, 1, 2 (broadcasted)

darosior replied:
Done

danielabrozzoni commented on src/daemon/control.rs:
```diff
@@ -456,6 +543,80 @@ fn share_unvault_signatures(
+                res_tx.as_psbt_string(),
+                spend_tx.as_psbt_string()
+            );
+            return Ok(());
```
Can you please explain why you're returning Ok(()) even if the cosigner server lied?

darosior replied:
To be honest i don't think we can have decent detailed errors until we get rid of the JSONRPC thread and stuff calls in jsonrpc/api.rs

This is just going to fail the finalization later, i know it's horrible..

darosior replied:
i mean i can just abuse jsonrpc::Error again..

darosior replied:
Oh, actually it can not happen https://github.com/revault/revault_tx/blob/b0502d50ac20d65fbc5f4960d3aae7f7e736da69/src/transactions.rs#L480-L487

darosior replied:
```diff
diff --git a/src/daemon/control.rs b/src/daemon/control.rs
index b377788..84603ba 100644
--- a/src/daemon/control.rs
+++ b/src/daemon/control.rs
@@ -579,22 +579,15 @@ fn fetch_cosigner_signatures(
                 one of these vaults!"
                 .to_string(),
         ))?;
-        if res_tx.inner_tx().inputs.len() != spend_tx.inner_tx().inputs.len() {
-            // They lied to us, no need to go further.
-            log::error!(
-                "Cosigning server sent us a bad Spend transaction: '{}', our is '{}'",
-                res_tx.as_psbt_string(),
-                spend_tx.as_psbt_string()
-            );
-            return Ok(());
-        }
 
         for (i, psbtin) in res_tx.into_psbt().inputs.into_iter().enumerate() {
             spend_tx
                 .inner_tx_mut()
                 .inputs
                 .get_mut(i)
-                .expect("Checked above")
+                .expect(
+                    "A SpendTransaction cannot have a different number of txins and PSBT inputs",
+                )
                 .partial_sigs
                 .extend(psbtin.partial_sigs);
         }
```

edouardparis commented on doc/API.md:
```diff
@@ -278,14 +289,73 @@ feerate.
+| -------------- | ------ | -------------------------------------------------------------------- |
+| `spend_txs`    | array  | Array of [Spend transaction resources](#spend_transaction_reources)  |
+
+##### Spend transaction resources
+
+| Field                  | Type          | Description                                                           |
+| ---------------------- | ------------- | --------------------------------------------------------------------- |
+| `deposit_outpoints`    | string array  | Array of the deposit outpoints of the vaults this transaction spends  |
+| `psbt`                 | string        | Base64-encoded Spend transaction PSBT                                 |
```
Is it possible to modify `getspendtx` response in this PR, in order to have the same `Spend transaction resource` for both `listspendtxs` and `getspendtx` ?

darosior replied:
Discussed in face to face and we agreed it was ugly to return something you've been input'ed

darosior commented:
Ok, pushed a hopefully final version, thanks everyone for bearing with me and my hacks

darosior commented:
Removed the concurrency too and opened https://github.com/revault/revaultd/issues/135

edouardparis reviewed (approved):
ACK a5ffd9c3449ee0798d4c98ad07bfae8669ce558c

danielabrozzoni commented on doc/API.md:
```diff
@@ -16,7 +16,12 @@ Note that all addresses are bech32-encoded *version 0* native Segwit `scriptPubK
 | [`revocationtxs`](#revocationtxs)                           | Give back the revocation transactions signed         |
 | [`unvaulttx`](#unvaulttx)                                   | Give back the unvault transaction signed             |
-| [`spendtx`](#spendtx)                                       | Give back the spend transaction signed               |
+| [`updatependtx`](#updatespendtx)                            | Store or update the stored Spend transaction         |
```
An s is missing ("update pend tx")

danielabrozzoni commented on doc/API.md:
```diff
@@ -278,14 +289,73 @@ feerate.
 
 | Field        | Type   | Description                                    |
 | ------------ | ------ | ---------------------------------------------- |
 | `spend_tx`   | string | Base64-encoded Spend transaction PSBT          |
```
I think `setspendtx` takes a txid?

darosior replied:
:sob:
Done

edouardparis reviewed (approved):
ACK bf3d941561396aba461274cc3d9e4f89f7f56cb0
